authorization/rbac/rbac.go

```diff
@@ -0,0 +1,112 @@
+package rbac
+
+import (
+	"path"
+
+	"github.com/grepory/storage/apis/rbac"
+	"github.com/grepory/storage/authorization"
+	"github.com/grepory/storage/storage"
+)
+
+// Authorizer implements an authorizer interface using Roles Base Acccess
+// Control (RBAC)
+type Authorizer struct {
+	store storage.Store
+}
+
+// Authorize determines if a request is authorized based on its attributes
+func (a *Authorizer) Authorize(attrs authorization.RequestAttributes) (bool, error) {
+	// Get cluster roles binding
+	clusterRoleBindings := []rbac.ClusterRoleBinding{}
+	if err := a.store.List("clusterrolebindings", &clusterRoleBindings); err != nil {
+		return false, err
+	}
+
+	// Inspect each cluster role binding
+	for _, clusterRoleBinding := range clusterRoleBindings {
+		// Verify if this cluster role binding matches our user
+		if !matchesUser(attrs.User, clusterRoleBinding.Subjects) {
+			continue
+		}
+
+		// Get the cluster role that matched our user
+		clusterRole := &rbac.ClusterRole{}
+		key := path.Join("clusterroles", clusterRoleBinding.RoleRef.Name)
+		if err := a.store.Get(key, clusterRole); err != nil {
+			return false, err
+		}
+
+		// Loop through the cluster role rules
+		for _, rule := range clusterRole.Rules {
+			// Verify if this rule applies to our request
+			if ruleAllows(attrs, rule) {
+				return true, nil
+			}
+		}
```

Comment:
Not exactly related to this code, but how do we deal with Rules that might be contradictory? Will we just make sure such contradictions in rules are impossible to create in the first place (by doing more work during creation) and/or should be still traverse the whole rule set here and detect potential contradictions (and return an appropriate error)?

Same comment/question applies to any traversal of `Rules` below.

Reply:
Hmm.. That's a good point. I would say that a reasonable behavior is "first match wins, watch out for the footgun."

Reply:
Wait. Contradictory rules aren't possible, as rules explicitly state what _is_ authorized. So it makes sense that the whole policy is the union of all rules. If one rule is less permissive than another and both apply to the same object, the more permissive rule should win.

Reply:
Exactly. From [K8s docs](https://kubernetes.io/docs/reference/access-authn-authz/authorization/):

> Kubernetes evaluates all of the request attributes against all policies and allows or denies the request. All parts of an API request must be allowed by some policy in order to proceed. This means that permissions are denied by default.

This behaviour really simplifies the rules compared to what we have so far in 2.x!

Reply:
Thanks for the clarification guys, I see no problem if permissions are indeed denied by default!